[Linalg] Add a runtime switch to let pytorch prefer a backend impl in linalg functions on GPU

xwang233 opened the pull request:
Per title.

This PR introduces a global flag that lets pytorch prefer one of the many backend implementations while calling linear algebra functions on GPU.

Usage:
```python
torch.backends.cuda.preferred_linalg_library('cusolver')
```

Available options (str): `'default'`, `'cusolver'`, `'magma'`.

Issue https://github.com/pytorch/pytorch/issues/63992 inspired me to write this PR. No heuristic is perfect on all devices, library versions, matrix shapes, workloads, etc. We can obtain better performance if we can conveniently switch linear algebra backends at runtime.

Performance of linear algebra operators after this PR should be no worse than before. The flag is set to **`'default'`** by default, which makes everything the same as before this PR.

The implementation of this PR is basically following that of https://github.com/pytorch/pytorch/pull/67790.

xwang233 committed:
linalg.cuda_prefer_cusolver flag py/c++ bindings

xwang233 committed:
global flag override heuristics

xwang233 committed:
doc and warning

xwang233 committed:
format

xwang233 commented:
cc @ptrblck

Lezcano commented:
I like this idea! Now, my only question is: should we also add a similar one for MAGMA, as otherwise it feels a bit lacking. 
It also makes testing different backends waaay easier, which is nice.

I think these are good points to discuss in tomorrow's meeting.

xwang233 commented:
@Lezcano Thanks for the comments. I have a similar idea to also add a flag for MAGMA, but there are some technical reasons. `ATen/Context.h` is a heavily included header, and most global flags are primitive types, specifically `bool`. I was thinking to add a enum type to a standalone file and include it in the `Context.h` header, but I'm not sure if that's the right approach. Yeah, let's discuss these technical details as well.

mruberry commented:
Notes from linalg meeting: 

- look at similar switch changing behavior of reductions
- a similar option to prefer MAGMA would be interesting
- we could consider adding more algo/driver arguments to ops
- this would let users override heuristics
- this would facilitate debugging/development
- let's document which functions can be affected by this flag?

Follow-ups:
- on whether this should be an enum now to allow for MAGMA expansion without a BC-breaking change.
- Loss reductions probably live in ATen and could be a good model to follow.
- Name needs bikeshedding. "preferred" sounds right

xwang233 committed:
Merge remote-tracking branch 'upstream/master' into linalg-runtime-switch

xwang233 committed:
LinalgBackend py/cpp bindings

xwang233 committed:
remove cuda_prefer_cusolver

xwang233 committed:
heuristic and preferred_backend

xwang233 committed:
Merge remote-tracking branch 'upstream/master' into linalg-runtime-switch

xwang233 committed:
doc

xwang233 committed:
lint

xwang233 committed:
mypy

xwang233 committed:
flake8

xwang233 committed:
warning

xwang233 committed:
clang format

xwang233 committed:
typo

xwang233 committed:
add a test

xwang233 committed:
skipCUDAIfRocm

IvanYashchuk commented on test/test_linalg.py:
```diff
@@ -8082,6 +8082,26 @@ def test_tensordot(self, device):
         self.assertEqual(a, an)
 
+    @onlyCUDA
+    @skipCUDAIfRocm
```
Is this test failing for ROCm? Sure "cuSOLVER" backend is not available there, but in this case, it should probably pick the "default" backend, no?

xwang233 replied:
Actually, yes, you're right. 😅  It'll pick the default magma path. Let me remove this rocm skip.

xwang233 committed:
override and binding

xwang233 committed:
doc coverage fix

You added the following module(s) to the PyTorch namespace 'torch.backends.linalg' but they have no corresponding entry in a doc .rst file. You should either make sure that the .rst file that contains the module's documentation properly contains either '.. automodule:: mod_name' (if you do not want the paragraph added by the automodule, you can simply use py:module) or make the module private (by appending an '_' at the beginning of its name.

xwang233 committed:
Merge remote-tracking branch 'upstream/master' into linalg-runtime-switch

xwang233 committed:
remove rocm skip

xwang233 commented:
Hi @ngimel , @mruberry , this PR is ready for review.

mruberry commented on aten/src/ATen/Context.cpp:
```diff
@@ -147,6 +147,21 @@ void Context::setAllowTF32CuBLAS(bool b) {
+    TORCH_WARN_ONCE(
+      "torch.backends.linalg.preferred is an experimental feature. "
+      "If you see any error or regression when this flag is set, "
+      "you're encourged to file an issue on GitHub."
```
"encourged" -> "encouraged"

mruberry commented on aten/src/ATen/Context.cpp:
```diff
@@ -147,6 +147,21 @@ void Context::setAllowTF32CuBLAS(bool b) {
+  if (b != at::LinalgBackend::Default) {
+    TORCH_WARN_ONCE(
+      "torch.backends.linalg.preferred is an experimental feature. "
+      "If you see any error or regression when this flag is set, "
```
"," -> ""

mruberry commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cpp:
```diff
@@ -2064,12 +2126,12 @@ Tensor& orgqr_kernel_impl(Tensor& result, const Tensor& tau) {
-    TORCH_CHECK(false, "Calling torch.orgqr on a CUDA tensor requires compiling ",
-      "PyTorch with cuSOLVER. Please use PyTorch built with cuSOLVER support.");
-  #endif
+#if defined(USE_CUSOLVER)
```
Previous formatting was OK, too

mruberry commented on docs/source/backends.rst:
```diff
@@ -74,6 +75,43 @@ torch.backends.cudnn
+
+    .. warning:: This flag is experimental and subject to change.
+
+    A flag that lets pytorch prefer one of the many backend implementations when calling linear algebra functions on GPU.
```
"pytorch" -> "PyTorch"

alt. "When PyTorch runs a CUDA linear algebra operation it often uses the cuSOLVER or MAGMA libraries, and if both are available it decides which to use with a heuristic. This flag allows overriding those heuristics. If `torch.linalg_cusolver` is set then cuSOLVER will be used wherever possible. If `torch.linalg_magma` is set then MAGMA will be used wherever possible. If `torch.linalg_default` (the default) is set then the heuristics will be used if both cuSOLVER and MAGMA are available. 

This flag When multiple CUDA math libraries implement a linear algebra

mruberry commented on docs/source/backends.rst:
```diff
@@ -74,6 +75,43 @@ torch.backends.cudnn
+    * Use as a global flag, e.g. `torch.backends.linalg.preferred = torch.linalg_cusolver`
+    * Use the context manager, e.g. `with torch.backends.linalg.flags(preferred=torch.linalg_cusolver):`
+
+    Note: The usage of a backend is not guaranteed for all linear algebra operators even if that backend is set as preferred.
```
alt. "When a library is preferred other libraries may still be used if the preferred library doesn't implement the operation(s) called."

mruberry commented on docs/source/backends.rst:
```diff
@@ -74,6 +75,43 @@ torch.backends.cudnn
+    * Use the context manager, e.g. `with torch.backends.linalg.flags(preferred=torch.linalg_cusolver):`
+
+    Note: The usage of a backend is not guaranteed for all linear algebra operators even if that backend is set as preferred.
+    Explicitly setting a preferred backend may override existing pytorch linear algebra heuristics and achieve better performance.
```
"This flag may achieve better performance if PyTorch's heuristic library selection is incorrect for your application's inputs."

mruberry reviewed (approved):
Cool! I added a few doc suggestions inline.

Adding a new enum looks onerous. Maybe we should simplify that process...

Let's let @ngimel take a look, too

xwang233 committed:
doc and warning message updates

xwang233 committed:
Merge remote-tracking branch 'upstream/master' into linalg-runtime-switch

mruberry commented on aten/src/ATen/Context.cpp:
```diff
@@ -147,6 +147,21 @@ void Context::setAllowTF32CuBLAS(bool b) {
+  linalg_preferred_backend = b;
+  if (b != at::LinalgBackend::Default) {
+    TORCH_WARN_ONCE(
+      "torch.backends.linalg.preferred is an experimental feature. "
```
@ngimel suggests 

torch.backends.cuda.preferred_linalg_library

To capture this is CUDA only and that it's preferring a particular library

xwang233 replied:
The flag can't be changed to the first level under `torch.backends.cuda`, so it is changed to `torch.backends.cuda.linalg.preferred_library`

mruberry commented on c10/core/LinalgBackend.h:
```diff
@@ -0,0 +1,38 @@
+
+namespace c10 {
+
+enum class LinalgBackend : int8_t { Default, Cusolver, Magma };
```
@ngimel suggests that c10 shouldn't need to call this, so this file should be in ATen near where the code for the cudnn enabling logic lives

mruberry commented on torch/backends/linalg/__init__.py:
```diff
@@ -0,0 +1,43 @@
+    return orig_flags
+
+
+@contextmanager
```
@ngimel points out that because this setting is global we shouldn't provide a context manager for it -- the context manager wouldn't be respected when multithreading

mruberry commented on torch/csrc/LinalgBackend.cpp:
```diff
@@ -0,0 +1,81 @@
+#include <torch/csrc/LinalgBackend.h>
```
@ngimel Suggests that creating a new namespace from C++ like this shouldn't be necessary

mruberry replied:
(@ngimel suspects you can just put the necessary functions/options into Python and call whatever C++ you like)

xwang233 replied:
If we want an enum type in python, those py/c++ bindings are necessary. Cudnn flags, cublas.matmul flags or other similar flags all use primitive type -- boolean.

I'm basically following the implementation of `torch.memory_format` (`at::MemoryFormat`), which has an enum type in python as well.

albanD replied:
As mentioned below, we now have pybind. So you can do significantly simpler things than what we had to do for memory format: https://github.com/pytorch/pytorch/blob/dbbb02474b9ad8044d061caee9fb0916da93101b/torch/csrc/Module.cpp#L970-L990

mruberry commented on torch/csrc/jit/python/python_sugared_value.cpp:
```diff
@@ -1083,6 +1084,10 @@ std::shared_ptr<SugaredValue> toSugaredValue(
       auto memory_format = reinterpret_cast<THPMemoryFormat*>(obj.ptr());
       const auto v = static_cast<int64_t>(memory_format->memory_format);
       return toSimple(g.insertConstant(v, loc));
+    } else if (THPLinalgBackend_Check(obj.ptr())) {
```
Is this necessary for jit support? I don't think previous switches added code here

mruberry commented on torch/csrc/utils/linalg_backends.cpp:
```diff
@@ -0,0 +1,37 @@
+#include <torch/csrc/utils/linalg_backends.h>
```
What's the point of this file?

mruberry commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -35,6 +36,7 @@ static std::unordered_map<std::string, ParameterType> type_map = {
   {"ScalarType", ParameterType::SCALARTYPE},
   {"Layout", ParameterType::LAYOUT},
   {"MemoryFormat", ParameterType::MEMORY_FORMAT},
+  {"LinalgBackend", ParameterType::LINALG_BACKEND},
```
This also wasn't done for setting context like deterministic flags. @ngimel suspects if you follow the deterministic flag PRs more closely you can avoid a lot of this

mruberry commented:
Hey @xwang233! @ngimel and I had a chance to look together and we left some inline comments. See #67946 and Kurt's work on the deterministic flag for maybe a better, simpler guide to the process of adding a new enum and flag for it.

xwang233 committed:
Merge remote-tracking branch 'upstream/master' into linalg-runtime-switch

xwang233 committed:
move c10/core/LinalgBackend.h to ATen

xwang233 committed:
remove jit changes

xwang233 committed:
undef linalg macro

xwang233 committed:
revert python_arg_parser changes

xwang233 committed:
change py bindings to torch.backends.cuda.linalg.preferred_library

albanD commented on torch/csrc/utils/linalg_backends.cpp:
```diff
@@ -0,0 +1,40 @@
+  }
+
+void initializeLinalgBackends() {
+  auto torch_module = THPObjectPtr(PyImport_ImportModule("torch"));
```
Why not put this in `torch.backends.linalg`? It looks surprising to have these in `torch.`

albanD commented on torch/csrc/LinalgBackend.cpp:
```diff
@@ -0,0 +1,81 @@
+
+PyTypeObject THPLinalgBackendType = {
+  PyVarObject_HEAD_INIT(nullptr, 0)
+  "torch.linalg_backend",                       /* tp_name */
```
Why do you need this full fledge custom objects?
If you just want to bind your union to python, you can do https://github.com/pytorch/pytorch/blob/dbbb02474b9ad8044d061caee9fb0916da93101b/torch/csrc/Module.cpp#L970-L990

xwang233 replied:
Thanks! Didn't know this pybind usage. I'll try to use this to simplify the PR.

albanD commented on docs/source/backends.rst:
```diff
@@ -45,6 +45,34 @@ torch.backends.cuda
 
         Clears the cuFFT plan cache.
 
+.. attribute::  torch.backends.cuda.linalg.preferred_library
```
The PR description talks about `torch.backends.linalg.preferred` while this one is under `cuda`. Is that on purpose?

xwang233 replied:
It was suggested here https://github.com/pytorch/pytorch/pull/67980#discussion_r751801464 to use `cuda` namespace. I forgot to update the PR description after the code change.

albanD commented on test/test_public_bindings.py:
```diff
@@ -147,6 +147,7 @@ def test_no_new_bindings(self):
             "is_inference_mode_enabled",
             "JITException",
             "layout",
+            "linalg_backend",
```
This should be properly documented in an rst file if you add it here

xwang233 committed:
Merge remote-tracking branch 'upstream/master' into linalg-runtime-switch

xwang233 committed:
use pybind11 for py/c++ bindings

xwang233 committed:
Merge remote-tracking branch 'upstream/master' into linalg-runtime-switch

xwang233 committed:
lint

xwang233 committed:
mypy

albanD commented on test/test_linalg.py:
```diff
@@ -8188,6 +8188,27 @@ def test_tensordot(self, device):
+        # The main purpose of this test is to make sure these "backend" calls work normally without raising exceptions.
+        x = torch.randint(2, 5, (2, 4, 4), device='cuda', dtype=torch.double)
+
+        torch.backends.cuda.preferred_linalg_library('cusolver')
```
Is this sending a warning? If so, it can be nice to catch it to avoid spam.
Also we have something to make warn once always want and you can actually check that you do get the warning.

xwang233 replied:
It's using TORCH_WARN_ONCE in Context.cpp. It raises `Warning` instead of `UserWarning` and doesn't work well with `assertWarnsOnceRegex`.

```
[W Context.cpp:157] Warning: torch.backends.cuda.preferred_linalg_library is an experimental feature. If you see any error or regression when this flag is set please file an issue on GitHub. (function operator())
```

albanD commented on test/test_linalg.py:
```diff
@@ -8188,6 +8188,27 @@ def test_tensordot(self, device):
+        torch.backends.cuda.preferred_linalg_library('cusolver')
+        out1 = torch.linalg.inv(x)
+
+        torch.backends.cuda.preferred_linalg_library('magma')
```
nit: on failure, this test will not properly reset the preferred library. Do we care enough to have a try/finally around the whole test?

xwang233 replied:
done

albanD reviewed (approved):
The binding code is much simpler! Thanks for taking the time to update it!

xwang233 committed:
wrap test with a try-finally

xwang233 committed:
Merge remote-tracking branch 'upstream/master' into linalg-runtime-switch

ngimel commented on aten/src/ATen/Context.cpp:
```diff
@@ -147,6 +147,21 @@ void Context::setAllowTF32CuBLAS(bool b) {
+  if (b != at::LinalgBackend::Default) {
+    TORCH_WARN_ONCE(
+      "torch.backends.cuda.preferred_linalg_library is an experimental feature. "
+      "If you see any error or regression when this flag is set "
```
Perf regressions are expected if you pick the wrong backend, right? So the error message shouldn't suggest filing issue in this case. It could suggest filing an issue for errors or unexpected behaviors.

xwang233 replied:
Thanks for the suggestion. I've changed the warning message.

ngimel reviewed (approved):
Please fix the warning message and ping me when it's done.

xwang233 committed:
Merge remote-tracking branch 'upstream/master' into linalg-runtime-switch

xwang233 committed:
change warning message

xwang233 commented:
Hi @ngimel , I've fixed the warning message. The CI failures seem irrelevant. Should I rebase and trigger the CI again or is this ready to go?

ngimel commented:
That's fine, you don't have to rebase
